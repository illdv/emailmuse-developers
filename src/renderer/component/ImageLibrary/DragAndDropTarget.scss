.dnd-target {
  min-width: 400px;
  min-height: 400px;
  height: 100%;
  width: 100%;
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
}

.dnd-target__overlay {
  background-color: #FFF;
  opacity: 0;
  visibility: hidden;
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  transition: visibility, opacity 0.15s ease;
  outline: 2px dashed #4a4a4a;
  outline-offset: -0.7rem;
  z-index: 1000;
  pointer-events: none;
}

.dnd-target__overlay--can-drop {
  opacity: 0.85 !important;
  visibility: visible !important;
}

.dnd-target__overlay--drag-over {
  background-color: #DDD;
}
