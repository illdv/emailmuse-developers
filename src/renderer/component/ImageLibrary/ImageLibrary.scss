.image-library {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.image-library__container {
  width: 95%;
  height: 95%;
  position: relative;
  display: flex;
  justify-content: space-around;
  flex-wrap: wrap;
}

.image-library__upload-input {
  //padding-top: 1rem;
  position: absolute;
  top: -1000;
  left: -1000;
  opacity: 0;
  visibility: hidden;
}

