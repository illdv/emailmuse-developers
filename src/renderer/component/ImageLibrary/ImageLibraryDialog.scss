.image-library-dialog {
    position: relative;
}

.image-library-dialog__img {
    display: flex;
    margin: auto;
    height: 9rem;
    max-width: 100%;
    overflow: hidden;
    border-radius: 0.25rem;
}

.image-library-dialog__content {
    overflow-y: hidden !important;
}

.image-library-dialog__url-field:hover ~ .image-library-dialog__url-tooltip {
    visibility: visible;
}

.image-library-dialog__url-tooltip {
    visibility: hidden;
    width: 240px;
    background-color: #ececec;
    position: absolute;
    right: 0;
    bottom: -1.9rem;
    text-align: center;
    padding: 5px 0;
}