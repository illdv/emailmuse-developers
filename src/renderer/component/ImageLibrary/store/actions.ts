import * as constants from './constants';
import { createAction } from 'redux-actions';
import { IGetImagesResponse } from 'src/renderer/component/ImageLibrary/store/models';
import * as Status from 'src/renderer/common/PreloaderLayout/Status/constants';

export const getImagesRequest = (page = 1, name: string = '') => ({
  type: constants.GET_IMAGES_REQUEST,
  payload: { page, name},
  meta: { status: Status.LOADING },
});

export const getImagesSuccess = (response: IGetImagesResponse) => ({
  type: constants.GET_IMAGES_SUCCESS,
  payload: response,
  meta: { status: Status.LOADED },
});

export const getImagesFailure = createAction(constants.GET_IMAGES_FAILURE, null, () => ({ status: Status.FAILED }));

export const uploadImagesRequest = (files: File | File[]) => ({
  type: constants.UPLOAD_IMAGES_REQUEST,
  payload: [].concat(files),
  meta: { status: Status.LOADING },
});

export const uploadImagesSuccess =
  createAction(constants.UPLOAD_IMAGES_SUCCESS, null, () => ({ status: Status.LOADED }));

export const uploadImagesFailure =
  createAction(constants.UPLOAD_IMAGES_FAILURE, null, () => ({ status: Status.FAILED }));

export const updateImageRequest = (payload: { imageId: number, name: string }) => ({
  type: constants.UPDATE_IMAGE_REQUEST,
  payload,
  meta: { status: Status.LOADING },
});

export const updateImageSuccess =
  createAction(constants.UPDATE_IMAGE_SUCCESS, null, () => ({ status: Status.LOADED }));

export const updateImageFailure =
  createAction(constants.UPDATE_IMAGE_FAILURE, null, () => ({ status: Status.FAILED }));

export const deleteImagesRequest = (ids: number | number[]) => ({
  type: constants.DELETE_IMAGES_REQUEST,
  payload: [].concat(ids),
  meta: { status: Status.LOADING },
});

export const deleteImagesSuccess =
  createAction(constants.DELETE_IMAGES_SUCCESS, null, () => ({ status: Status.LOADED }));

export const deleteImagesFailure =
  createAction(constants.DELETE_IMAGES_FAILURE, null, () => ({ status: Status.FAILED }));
