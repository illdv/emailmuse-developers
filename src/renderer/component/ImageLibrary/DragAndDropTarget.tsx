import * as React from 'react';
import { DropTarget } from 'react-dnd';
import 'src/renderer/component/ImageLibrary/DragAndDropTarget.scss';
import block from 'bem-ts';
const b = block('dnd-target');

namespace DragAndDropTargetSpace {
  export interface IProps {
    onDrop: (item:any) => void;
    showOverlay?: boolean;
    overlayMessage?: string;
    connectDropTarget?: any;
    isOverCurrent?: boolean;
  }
  export interface IState {

  }
}

const types = {
  FILE: '__NATIVE_FILE__'
};

const specs = {
   drop(props, monitor, component) {
     if (props.onDrop) {
       props.onDrop(monitor.getItem());
     }
  }
};

const collect = (connect, monitor) => ({
  connectDropTarget: connect.dropTarget(),
  isOverCurrent: monitor.isOver({ shallow: true })
});

@DropTarget(types.FILE, specs, collect)
export class DragAndDropTarget extends React.Component<DragAndDropTargetSpace.IProps, DragAndDropTargetSpace.IState> {
  constructor (props) {
    super(props);
  }

  render() {
    const {connectDropTarget, isOverCurrent, showOverlay, overlayMessage, children} = this.props;
    return connectDropTarget(
      <div className={b()}>
        {showOverlay ? <div className={b('overlay-background', {dragging: isOverCurrent})}/> : null}
        {showOverlay && overlayMessage ?
          <div className={b('overlay-message', {dragging: isOverCurrent})}>
            <span>{overlayMessage}</span>
          </div> :
          null
        }
        {children}
      </div>
    );
  }
}
