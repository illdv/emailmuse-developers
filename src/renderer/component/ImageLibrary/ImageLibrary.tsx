import * as React from 'react';
import { IStyle } from 'type/materialUI';
import { Paper, withStyles } from '@material-ui/core/';
import { TypeBackground } from '@material-ui/core/styles/createPalette';
import { DragAndDropTarget } from './DragAndDropTarget';
import { ImageLibraryListComponent } from './ImageLibraryList';
import * as EmailerAPI from 'src/renderer/API/EmailerAPI';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { getImagesRequest } from 'src/renderer/component/ImageLibrary/store/actions';

import 'src/renderer/component/ImageLibrary/ImageLibrary.scss';
import block from 'bem-ts';
const b = block('image-library');

namespace ImageLibrarySpace {
  export interface IProps {
    classes?: any;
    actions?: {
      getImagesRequest: typeof getImagesRequest
    };
  }
  export interface IState {
    items: File[];
  }
}

const styles: IStyle = theme => ({
  root: {
    width: '100%',
    height: '100%',
    backgroundColor: theme.palette.background.paper,
  }
});

const mapStateToProps = state => ({
  // TODO refactor with selectors
  items: state.images.all
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ getImagesRequest }, dispatch)
});

@connect(mapStateToProps, mapDispatchToProps)
class ImageLibrary extends React.Component<ImageLibrarySpace.IProps, ImageLibrarySpace.IState> {
  constructor (props) {
    super(props);
    this.state = {
      items: []
    };
  }

  componentDidMount () {
    this.props.actions.getImagesRequest();
  }

  onDrop = item => {
    if (item && item.files) {
      this.addFiles(item.files);
    }
  }

  onProgress = (percentage) => {
    console.log('Loaded:', percentage);
  }

  // TODO: Should user be prevented from uploading one same image twice
  // and how two images could be identified as 'same'?
  addFiles = files => {
    const filtered = files.filter(file => !this.state.items.some(item => item.name === file.name));
    this.setState({items: [...this.state.items, ...filtered]});
    // tslint:disable-next-line
    // console.log('Files has been added', filtered);

    // const filo = filtered[0];
    // let fileData = new FormData();
    // fileData.append(filo.name, filo.path);

    // EmailerAPI.ImageLibrary.uploadImage(filtered[0], this.onProgress).then(result => {
    //   console.log('Post done', result);
    // });

    // EmailerAPI.ImageLibrary.getImages().then(result => {
    //   console.log('Get done', result);
    // });

    // EmailerAPI.ImageLibrary.updateImage(1, "lol").then(result => {
    //   console.log('Get done', result);
    // });

    // EmailerAPI.ImageLibrary.deleteImages([1,2]).then(result => {
    //   console.log('Get done', result);
    // });
    // filtered[0]
  }

  render() {
    const { classes } = this.props;
    return (
      <Paper elevation={4} className={classes.root}>
        <div className={b()}>
          <DragAndDropTarget
            onDrop={this.onDrop}
            showOverlay={true}
            overlayMessage={'Drop files here to add them to your image library'}
          >
            <div className={b('container')}>
              <ImageLibraryListComponent items={this.state.items}/>
            </div>
          </DragAndDropTarget>
        </div>
      </Paper>
    );
  }
}

export default withStyles(styles)(ImageLibrary as any);
