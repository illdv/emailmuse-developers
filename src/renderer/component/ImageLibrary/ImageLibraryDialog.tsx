import * as React from 'react';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import TextField from '@material-ui/core/TextField';
import InputAdornment from '@material-ui/core/InputAdornment';
import Language from '@material-ui/icons/Language';
import { IImageLibraryItem } from 'src/renderer/component/ImageLibrary/store/models';
import block from 'bem-ts';
const b = block('image-library-dialog');
import 'src/renderer/component/ImageLibrary/ImageLibraryDialog.scss';

namespace ImageLibraryDialogSpace {
  export interface IProps {
    item: IImageLibraryItem;
    onUpdateItem: (item:IImageLibraryItem, newName:string) => void;
    onDeleteItem: (item:IImageLibraryItem) => () => void;
    onClose: () => void;
  }
  export interface IState {
    newName: string;
  }
}

export class ImageLibraryDialog
  extends React.Component<ImageLibraryDialogSpace.IProps, ImageLibraryDialogSpace.IState> {
  constructor(props) {
    super(props);
    this.state = {
      newName: props.item.name
    };
  }

  handleDialogClose = () => {
    this.props.onClose();
  }

  handleUpdateItem = (e) => {
    e.preventDefault();
    this.props.onUpdateItem(this.props.item, this.state.newName);
    this.handleDialogClose();
  }

  handleInput = (e) => {
    this.setState({ newName: e.target.value });
  }

  handleDeleteItem = () => {
    this.props.onDeleteItem(this.props.item)();
    this.handleDialogClose();
  }

  inputClick = (e) => {
    e.preventDefault();
    const p = document.getElementById('url') as HTMLInputElement;
    p.select();
    document.execCommand('copy');
  }

  preventDefault = (e) => {
    e.preventDefault();
  }

  render() {
    return (
      <Dialog
        open={true}
        onClose={this.handleDialogClose}
        aria-labelledby="form-dialog-title"
      >
        <form onSubmit={this.handleUpdateItem}>
          <DialogTitle id="form-dialog-title">Edit image</DialogTitle>
          <DialogContent className={b('content')}>
            <DialogContentText>Enter new name of this image and click 'Update' to apply changes.</DialogContentText>
            <DialogContentText>Or click 'Delete' to delete this image from library.</DialogContentText>
            <img
              src={this.props.item.url}
              className={b('img')}
            />
            <TextField
              autoFocus
              margin="dense"
              id="name"
              label="Image name"
              type="input"
              value={this.state.newName}
              fullWidth
              onChange={this.handleInput}
            />
            <TextField
              margin="dense"
              id="url"
              label="Image URL"
              type="input"
              value={this.props.item.url}
              fullWidth
              onClick={this.inputClick}
              onChange={this.preventDefault}
              InputProps={{
                startAdornment: (
                  <InputAdornment
                    position="start"
                    className={b('url-field')}
                  >
                    <Language />
                  </InputAdornment>
                ),
              }}
            />
          </DialogContent>
          <DialogActions>
            <Button onClick={this.handleDialogClose} color="primary">
              Cancel
            </Button>
            <Button
              color="primary"
              type="submit"
            >
              Update
            </Button>
            <Button onClick={this.handleDeleteItem} color="primary">
              Delete
            </Button>
          </DialogActions>
        </form>
      </Dialog>
    );
  }
}