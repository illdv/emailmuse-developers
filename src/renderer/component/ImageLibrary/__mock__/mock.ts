import { IImageLibraryItem } from '../store/models';

export const mockImage0: IImageLibraryItem = {
  id: 89,
  url:
    'https://infocron.s3.us-east-2.amazonaws.com/user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg.jpeg',
  thumb_url:
    'https://infocron.s3.us-east-2.amazonaws.com/' +
    'user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg_thumb450.jpeg',
  name: 'tesla-cat',
};
export const mockImage1: IImageLibraryItem = {
  id: 90,
  url:
    'https://infocron.s3.us-east-2.amazonaws.com/user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg.jpeg',
  thumb_url:
    'https://infocron.s3.us-east-2.amazonaws.com/' +
    'user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg_thumb450.jpeg',
  name: 'tesla-cat',
};
export const mockImage2: IImageLibraryItem = {
  id: 91,
  url:
    'https://infocron.s3.us-east-2.amazonaws.com/user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg.jpeg',
  thumb_url:
    'https://infocron.s3.us-east-2.amazonaws.com/' +
    'user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg_thumb450.jpeg',
  name: 'tesla-cat',
};
export const mockImageArray: IImageLibraryItem[] = [
  mockImage0,
  mockImage1,
  mockImage2,
];
