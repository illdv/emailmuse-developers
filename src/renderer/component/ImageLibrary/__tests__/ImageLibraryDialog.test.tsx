import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { mockImage0 } from '../__mock__/mock';
import { ImageLibraryDialog } from 'src/renderer/component/ImageLibrary/ImageLibraryDialog';
import { ToastType } from 'src/renderer/common/Toast/flux/actions';

describe('<ImageLibraryDialog />', () => {
  let dialog;
  let props;
  const mockClose              = jest.fn();
  const mockUpdateItem         = jest.fn();
  const mockDeleteItemReturned = jest.fn();
  const mockDeleteItem         = jest.fn().mockReturnValue(() => mockDeleteItemReturned);
  const mockShowToast = jest.fn();

  beforeAll(() => {
    props  = {
      item: mockImage0,
      onUpdateItem: mockUpdateItem,
      onDeleteItem: mockDeleteItem,
      onClose: mockClose,
      onShowToast: mockShowToast,
    };
    dialog = shallow(<ImageLibraryDialog {...props} />);
  });

  describe('Rendering UI', () => {
    it('Renders Correctly', () => {
      expect(dialog).toMatchSnapshot();
    });
  });

  describe('User interact with dialog,', () => {
    const typedValue = 'New name';
    beforeAll(() => {
      props  = {
        item: mockImage0,
        onUpdateItem: mockUpdateItem,
        onDeleteItem: mockDeleteItem,
        onClose: mockClose,
        onShowToast: mockShowToast,
      };
      dialog = shallow(<ImageLibraryDialog {...props} />);
    });

    it('he click on the inputName and types text', () => {
      dialog
        .find('.image-library-dialogs__inputName')
        .simulate('change', { target: { value: typedValue } });

      expect(dialog.state().newName).toBe(typedValue);
    });

    it('he submit the form', () => {
      dialog
        .find('.image-library-dialogs__inputName')
        .simulate('change', { target: { value: typedValue } });
      dialog.find('form')
        .simulate('submit', { preventDefault() {} });

      expect(mockUpdateItem).toBeCalledWith(props.item, typedValue);
    });

    describe('he click on the inputURL', () => {
      let mountedDialog;
      beforeEach(() => {
        Object.defineProperty(window.document, 'execCommand', {
          writable: true,
          value: jest.fn().mockReturnValueOnce(true).mockReturnValueOnce(false),
        });
        props         = {
          item: mockImage0,
          onUpdateItem: mockUpdateItem,
          onDeleteItem: mockDeleteItem,
          onClose: mockClose,
          onShowToast: mockShowToast,
        };
        mountedDialog = mount(<ImageLibraryDialog {...props} />);
        mountedDialog
          .find('.image-library-dialogs__inputURL')
          .first()
          .simulate('click', { preventDefault() {} })
          .simulate('click', { preventDefault() {} });
      });
      afterAll(() => {
        mountedDialog.unmount();
      });

      test('URL copied to clipboard', () => {
        expect(mockShowToast.mock.calls[0][1]).toBe(ToastType.Success);
      });

      test('URL did not copied to clipboard', () => {
        expect(mockShowToast.mock.calls[1][1]).toBe(ToastType.Error);
      });
    });

    it('he click delete button', () => {
      dialog.find('.image-library-dialogs__button--delete').simulate('click');

      expect(mockDeleteItem).toBeCalledWith(mockImage0);
      expect(mockClose).toBeCalled();
    });

    it('he close up the form', () => {
      dialog.instance().handleDialogClose();

      expect(mockClose).toBeCalled();
    });
  });

});
