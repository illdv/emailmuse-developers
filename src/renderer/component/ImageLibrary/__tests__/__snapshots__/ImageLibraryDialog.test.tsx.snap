// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageLibraryDialog /> Rendering UI Renders Correctly 1`] = `
<WithStyles(Dialog)
  aria-labelledby="form-dialog-title"
  maxWidth={false}
  onClose={[Function]}
  open={true}
>
  <form
    onSubmit={[Function]}
  >
    <WithStyles(DialogTitle)
      id="form-dialog-title"
    >
      Edit image
    </WithStyles(DialogTitle)>
    <WithStyles(DialogContent)
      className="image-library-dialogs__content"
    >
      <WithStyles(DialogContentText)>
        Enter new name of this image and click 'Update' to apply changes.
      </WithStyles(DialogContentText)>
      <WithStyles(DialogContentText)>
        Or click 'Delete' to delete this image from library.
      </WithStyles(DialogContentText)>
      <img
        className="image-library-dialogs__img"
        src="https://infocron.s3.us-east-2.amazonaws.com/user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg.jpeg"
      />
      <TextField
        autoFocus={true}
        className="image-library-dialogs__inputName"
        fullWidth={true}
        id="name"
        label="Image name"
        margin="dense"
        onChange={[Function]}
        required={false}
        select={false}
        type="input"
        value="tesla-cat"
      />
      <TextField
        InputProps={
          Object {
            "startAdornment": <WithStyles(InputAdornment)
              className="image-library-dialogs__url-field"
              position="start"
            >
              <pure(Language) />
            </WithStyles(InputAdornment)>,
          }
        }
        className="image-library-dialogs__inputURL"
        fullWidth={true}
        id="url"
        label="Image URL"
        margin="dense"
        onClick={[Function]}
        required={false}
        select={false}
        type="input"
        value="https://infocron.s3.us-east-2.amazonaws.com/user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg.jpeg"
      />
    </WithStyles(DialogContent)>
    <WithStyles(DialogActions)>
      <WithStyles(Button)
        className="image-library-dialogs__button--cancel"
        color="primary"
        onClick={[Function]}
      >
        Cancel
      </WithStyles(Button)>
      <WithStyles(Button)
        className="image-library-dialogs__button--update"
        color="primary"
        type="submit"
      >
        Update
      </WithStyles(Button)>
      <WithStyles(Button)
        className="image-library-dialogs__button--delete"
        color="primary"
        onClick={[Function]}
      >
        Delete
      </WithStyles(Button)>
    </WithStyles(DialogActions)>
  </form>
</WithStyles(Dialog)>
`;
